File: programmers/ForKaKAO/2020/level2/괄호 변환.cpp
#include <string>
#include <vector>
#include <stack>
#include <iostream>

using namespace std;

int isCorrect(string p, int *idx) {
    stack<char> st;
    int lp = 0, rp = 0;
    bool ret = true;
    for (int i = 0; i < p.length(); i++) {
        if (p[i] == '(') {
            st.push('(');
            lp++;
        } else {
            if (st.empty()) {
                ret = false;
            } else {
                st.pop();
            }
            rp++;
        }
        if (lp == rp) {
            *idx = i + 1;
            return ret;
        }
    }
}

string solution(string p) {
    if (p.empty()) {
        return p;
    }
    int idx = 0;
    bool flag = isCorrect(p, &idx);
    string u = p.substr(0, idx);
    string v = p.substr(idx);
    string str = "";
    
    if (flag) {
        return u + solution(v);
    } else {
        str = "(" + solution(v) + ")";
        
        for (int i = 1; i < u.length() - 1; i++) {
            if (u[i] == '(') {
                str += ")";
            } else {
                str += "(";
            }
        }
    }
    return str;
}

File: programmers/ForKaKAO/2020/level2/괄호 변환.py
def isCorrect(s):
    left = right = 0
    flag = True
    st = []
    
    for i in range(len(s)):
        ch = s[i]
        
        if ch == '(':
            left += 1
            st.append('(')
        else:
            right += 1
            if len(st) == 0:
                flag = False
            else:
                st.pop()
        
        if left == right:
            return i + 1, flag
        
    return 0, False
    
    
def solution(p):    
    if len(p) == 0:
        return p
    
    pos, flag = isCorrect(p)
    u = p[:pos]
    v = p[pos:]
    
    if flag:
        return u + solution(v)
    else:
        answer = '(' + solution(v) + ')'
        
        for i in range(1, len(u) - 1):
            if u[i] == '(':
                answer += ')'
            else:
                answer += '('
        
    return answer